analyzers/src/SonarAnalyzer.CSharp/Rules/PasswordsShouldBeStoredCorrectly.cs

```diff
@@ -0,0 +1,137 @@
+                        ctor => ctor.ContainingType.Is(KnownType.System_Security_Cryptography_Rfc2898DeriveBytes) && ctor.Parameters.Any(x => x.Name == "hashAlgorithm"),
+                        (methodName, comparison) => string.Compare(methodName, "Rfc2898DeriveBytes", comparison) == 0,
+                        null,
+                        x => x.Name == "iterations",
+                        null,
+                        null);
+        var invocationArgument = ArgumentDescriptor.MethodInvocation(KnownType.System_Security_Cryptography_Rfc2898DeriveBytes, "Pbkdf2", "iterations", x => x is 2 or 3);
+        Track(
+            argumentTracker,
+            context,
+            UseMoreIterationsMessageFormat,
+            argumentTracker.Or(
+                argumentTracker.MatchArgument(constructorArgument),
+                argumentTracker.MatchArgument(invocationArgument)),
+            x => ArgumentLessThan(x, IterationCountThreshold));
+
+        // Raise when hashAlgorithm is NOT present
+        var objectCreationTracker = CSharpFacade.Instance.Tracker.ObjectCreation;
+        Track(
+            objectCreationTracker,
+            context,
+            "Use at least 100 000 iterations and a state-of-the-art digest algorithm here.",
+            objectCreationTracker.MatchConstructor(KnownType.System_Security_Cryptography_Rfc2898DeriveBytes),
+            x => x.InvokedConstructorSymbol.Value.Parameters.All(x => x.Name != "hashAlgorithm"));
+    }
+
+    private static void BouncyCastle(SonarAnalysisContext context)
+    {
+        var tracker = CSharpFacade.Instance.Tracker.Argument;
+
+        Track(
+            tracker,
+            context,
+            "Use a cost factor of at least 12 here.",
+            tracker.Or(
+                tracker.MatchArgument(ArgumentDescriptor.MethodInvocation(KnownType.Org_BouncyCastle_Crypto_Generators_OpenBsdBCrypt, "Generate", "cost", x => x is 2 or 3)),
+                tracker.MatchArgument(ArgumentDescriptor.MethodInvocation(KnownType.Org_BouncyCastle_Crypto_Generators_BCrypt, "Generate", "cost", 2))),
+            x => ArgumentLessThan(x, 12));
+
+        Track(
+            tracker,
+            context,
+            UseMoreIterationsMessageFormat,
+            tracker.MatchArgument(ArgumentDescriptor.MethodInvocation(KnownType.Org_BouncyCastle_Crypto_PbeParametersGenerator, "Init", "iterationCount", 2)),
+            x => ArgumentLessThan(x, IterationCountThreshold));
+    }
+
+    private static bool HasFewIterations(PropertyAccessContext context, PropertyAccessTracker<SyntaxKind> tracker) =>
+        tracker.AssignedValue(context) is int iterationCount
+        && iterationCount < IterationCountThreshold;
+
+    private static bool ArgumentLessThan(ArgumentContext context, int threshold) =>
+        context.SemanticModel.GetConstantValue(((ArgumentSyntax)context.Node).Expression) is { HasValue: true, Value: int value }
+        && value < threshold;
```

Comment:
I reuse this for iterations and cost, and will re-use it once more for BouncyCastle's SCrypt.
I could make a wrapper for iterations on arguments, like the one I used for properties, but I think it's pretty readable already.
Let me know if you disagree.